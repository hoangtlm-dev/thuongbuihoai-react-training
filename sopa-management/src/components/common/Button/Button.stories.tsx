import type { Meta, StoryObj } from '@storybook/react';
import Button, { ButtonType } from '.';

const meta: Meta<typeof Button> = {
  component: Button,
};

export default meta;
type Story = StoryObj<typeof Button>;

export const ButtonPrimary: Story = {
  args: {
    text: "SHOW MORE",
    type: ButtonType.primary,
  }
};

export const ButtonSecondary: Story = {
  args: {
    text: "Add to bag $15",
    type: ButtonType.secondary,
  }
};

export const ButtonTertiary: Story = {
  args: {
    text: "LOGIN",
    type: ButtonType.tertiary,
  }
};

export const ButtonIconPrimary: Story = {
  args: {
    text: "",
    type: ButtonType.btnIconPrimary,
    icon: true
  }
};

export const ButtonIconSecondary: Story = {
  args: {
    text: "",
    type: ButtonType.btnIconSecondary,
    icon: true
  }
};

export const ButtonDisable: Story = {
  args: {
    text: "SHOW MORE",
    disable: true
  }
};
