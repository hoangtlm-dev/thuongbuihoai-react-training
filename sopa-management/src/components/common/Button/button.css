.btn-primary {
  padding: var(--px-normal) 0;
  width: 181px;
  background-color: var(--text-primary);
  border-radius:  var(--rounded-primary);
  border: none;
}

.btn-secondary,
.btn-tertiary {
  background-color: var(--dark);
  border-radius: var(--rounded-primary);
  color: var(--light);
  padding: var(--px-normal) 0;
  width: var(--w-medium);
}

.btn-tertiary {
  width: var(--w-large);
  padding: var(--px-normal) 0;
}

.btn-icon-primary {
  background-color: transparent;
  border: none;
}

.btn-icon-secondary {
  background-color: transparent;
  border-radius: 50%;
  border: 1px solid var(--border-primary);
  width: var(--w-medium);
  padding: var(--p-tiny);
}
