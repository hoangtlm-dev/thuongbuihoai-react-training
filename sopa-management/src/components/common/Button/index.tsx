import React, { ButtonHTMLAttributes, memo } from 'react';

// Styles
import './button.css';

export enum ButtonType {
  default = '',
  primary = 'primary',
  secondary = 'secondary',
  tertiary = 'tertiary',
  btnIconPrimary = 'icon-primary',
  btnIconSecondary = 'icon-secondary'
}

interface ButtonProps {
  children?: string | JSX.Element;
  type?: ButtonType;
  disable?: boolean;
  submit?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
  className?: string;
  onClick?: () => void;
  onSubmit?: () => void;
}

const Button: React.FC<ButtonProps> = ({
  children = '',
  type = ButtonType.default,
  disable = false,
  submit,
  className = '',
  onClick = () => {},
  onSubmit = () => {}
}) => (
  <button
    data-testid='button'
    className={`${className} btn-${type}`}
    disabled={disable}
    type={submit}
    onClick={onClick}
    onSubmit={onSubmit}
  >
    {children}
  </button>
);

export default memo(Button);
