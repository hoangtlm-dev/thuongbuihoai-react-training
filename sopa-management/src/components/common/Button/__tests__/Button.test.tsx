import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import Button, { ButtonType } from "../index";

describe("Button component", () => {
  test("Should render Button component with type 'default'", async () => {
    const { getByTestId } = render(<Button icon text="" type={ButtonType.default} />);
    const button = getByTestId("button");
    fireEvent.click(button)
    expect(button).toBeInTheDocument();
  });

  test("Should render Button component with type 'disable'", async () => {
    render(<Button icon disable text="" type={ButtonType.default} />);
    const clickBtn = screen.getByRole("button");
    expect(clickBtn).toBeInTheDocument();
  });

  test("Should render Button component with type 'primary'", async () => {
    render(<Button text="SHOW MORE" type={ButtonType.primary} />);

    const clickBtn = screen.getByRole("button", { name: /SHOW MORE/i });
    expect(clickBtn).toBeInTheDocument();
  });

  test("Should render Button component with type 'secondary'", async () => {
    render(<Button text="Add to bag" type={ButtonType.secondary} />);

    const clickBtn = screen.getByRole("button", { name: /Add to bag/i });
    expect(clickBtn).toBeInTheDocument();
  });

  test("Should render Button component with type 'tertiary'", async () => {
    render(<Button text="Add to bag" type={ButtonType.tertiary} />);

    const clickBtn = screen.getByRole("button", { name: /Add to bag/i });
    expect(clickBtn).toBeInTheDocument();
  });

  test("Should render Button component with type 'btnIconPrimary'", async () => {
    render(<Button icon type={ButtonType.btnIconPrimary} />);

    const clickBtn = screen.getByRole("button");
    expect(clickBtn).toBeInTheDocument();
  });

  test("Should render Button component with type 'btnIconSecondary'", async () => {
    render(<Button icon type={ButtonType.btnIconSecondary} />);

    const clickBtn = screen.getByRole("button");
    expect(clickBtn).toBeInTheDocument();
  });

  test("Should render Button component with default value", async () => {
    render(<Button />);

    const clickBtn = screen.getByRole("button");
    expect(clickBtn).toBeInTheDocument();
  });

  test("Should render Button component with action onClick", async () => {
    const onClickButton = jest.fn();
    render(
      <Button
        text="SHOW MORE"
        onClick={onClickButton}
        type={ButtonType.secondary}
      />,
    );

    const clickBtn = screen.getByRole("button", { name: /SHOW MORE/i });
    fireEvent.click(clickBtn);
    expect(onClickButton).toHaveBeenCalled();
  });

  test("Matches snapshot", () => {
    const { container } = render(
      <Button text="SHOW MORE" type={ButtonType.default} />,
    );
    expect(container).toMatchSnapshot();
  });
});
