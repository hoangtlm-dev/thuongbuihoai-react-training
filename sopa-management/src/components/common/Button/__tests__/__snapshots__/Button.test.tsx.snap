// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Matches snapshot 1`] = `
<div>
  <button
    class=" btn-"
    data-testid="button"
  >
    SHOW MORE
  </button>
</div>
`;
